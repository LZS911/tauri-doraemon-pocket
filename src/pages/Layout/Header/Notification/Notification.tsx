import { BellOutlined, CheckCircleOutlined } from '@ant-design/icons';
import { Popover, Tooltip } from 'antd';
import { useEffect, useMemo, useState } from 'react';
import { useTranslation } from 'react-i18next';
import {
  genTypeIcon,
  INotificationList,
  mockNotificationList,
  NotificationStatusEnum,
} from '.';
import ThemeBase from '../../../../components/ThemeBase';

const Notification: React.FC = () => {
  const { t } = useTranslation();
  const [notificationList, setNotificationList] = useState<INotificationList[]>(
    []
  );

  const notificationTipsCount = useMemo(() => {
    return notificationList.filter(
      (v) => v.status === NotificationStatusEnum.unread
    )?.length;
  }, [notificationList]);

  const markAsAllRead = () => {
    setNotificationList((list) => {
      const cloneList = [...list];
      list.forEach((v) => {
        v.status = NotificationStatusEnum.read;
      });
      return cloneList;
    });
  };

  useEffect(() => {
    mockNotificationList().then((res) => {
      setNotificationList(res);
    });
  }, []);

  return (
    <Popover
      content={
        <div className="w-[26rem]">
          <header className="flex items-center justify-between p-5 dark:pt-2">
            <span className="text-sm">
              {t('layout.header.notification.title')}
            </span>
            <ThemeBase.Icon
              data-testid="mark-all-read"
              hidden={notificationTipsCount === 0}
              onClick={markAsAllRead}
              icon={
                <CheckCircleOutlined className="text-[1rem] text-primary" />
              }
              tooltip={{
                title: t('layout.header.notification.titleTips'),
              }}
              className="hover:!bg-secondary"
            />
          </header>
          <div>
            {notificationList.map((v) => {
              return (
                <div
                  className={`flex cursor-pointer items-center border-t border-stone-100 px-4 py-2  dark:border-stone-700 ${
                    v.status === NotificationStatusEnum.read
                      ? 'hover:bg-zinc-200 dark:hover:bg-zinc-700'
                      : 'bg-slate-50  hover:!bg-secondary dark:bg-black'
                  }`}
                  key={v.id}
                >
                  <div>{genTypeIcon(v.type)}</div>
                  <div className="mx-4 w-[16rem]">
                    <Tooltip title={v.message}>
                      <div className="mb-1 truncate">
                        <span className="text-sm">{v.message}</span>
                      </div>
                    </Tooltip>

                    <div className="text-xs opacity-50">{v.desc}</div>
                  </div>
                  <div className="text-xs">{v.time}</div>
                </div>
              );
            })}
          </div>
          <footer className="w-full cursor-pointer border-t border-stone-100 px-4 py-3 text-center text-primary transition-[background] hover:bg-slate-200 dark:border-stone-700 dark:hover:bg-zinc-700 ">
            {t('layout.header.notification.viewAll')}
          </footer>
        </div>
      }
      trigger="click"
    >
      <ThemeBase.Icon
        data-testid="popover-switch-icon"
        badge={{
          count: notificationTipsCount,
        }}
        icon={<BellOutlined className="text-[1rem]" />}
        className="mr-2 bg-slate-100 dark:bg-black"
      />
    </Popover>
  );
};

export default Notification;
