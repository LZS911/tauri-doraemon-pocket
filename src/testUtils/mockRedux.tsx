import { configureStore } from '@reduxjs/toolkit';
import { combineReducers } from 'redux';
import layout from '../store/layout';
import config from '../store/config';
import { Dictionary } from '../typing/common.type';
import { Provider } from 'react-redux';
import React, { ReactNode } from 'react';
const reducers = combineReducers({
  layout,
  config,
});
export const storeFactory: any = (initStore: Dictionary = {}) => {
  return configureStore({
    reducer: reducers,
    preloadedState: initStore,
  });
};

export const CustomProvider: React.FC<{
  initStore: Dictionary;
  children: ReactNode;
}> = (props) => {
  return (
    <Provider store={storeFactory(props.initStore)}>{props.children}</Provider>
  );
};
