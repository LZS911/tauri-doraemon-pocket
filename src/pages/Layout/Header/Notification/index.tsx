import {
  CloseCircleOutlined,
  ExclamationCircleOutlined,
  SettingOutlined,
  WarningOutlined,
} from '@ant-design/icons';
import i18n from '../../../../locale';
import Notification from './Notification';

export default Notification;

export enum NotificationTypeEnum {
  system = 'system',
  information = 'information',
  error = 'error',
  warn = 'warn',
}
export enum NotificationStatusEnum {
  read = 'read',
  unread = 'unread',
}
export interface INotificationList {
  id: number;
  status: NotificationStatusEnum;
  message: string;
  desc: string;
  time: string;
  type: NotificationTypeEnum;
}

export const mockNotificationList = () => {
  return Promise.resolve<INotificationList[]>([
    {
      id: 1,
      status: NotificationStatusEnum.unread,
      message: i18n.t('backend.notification.message1'),
      desc: i18n.t('backend.notification.desc1'),
      time: '8:00 AM',
      type: NotificationTypeEnum.system,
    },
    {
      id: 2,
      status: NotificationStatusEnum.unread,
      message: i18n.t('backend.notification.message2'),
      desc: i18n.t('backend.notification.desc2'),
      time: '10:00 AM',
      type: NotificationTypeEnum.information,
    },
    {
      id: 3,
      status: NotificationStatusEnum.read,
      message: i18n.t('backend.notification.message3'),
      desc: i18n.t('backend.notification.desc3'),
      time: '1:00 PM',
      type: NotificationTypeEnum.error,
    },
    {
      id: 4,
      status: NotificationStatusEnum.read,
      message: i18n.t('backend.notification.message4'),
      desc: i18n.t('backend.notification.desc4'),
      time: '4:00 PM',
      type: NotificationTypeEnum.warn,
    },
  ]);
};

export const genTypeIcon = (type: NotificationTypeEnum) => {
  const typeCorrespondIconMap = new Map<NotificationTypeEnum, JSX.Element>([
    [
      NotificationTypeEnum.system,
      <div
        className="flex h-7 w-7 items-center justify-center rounded-[50%] bg-violet-50"
        key="system"
      >
        <SettingOutlined className="text-violet-500" />
      </div>,
    ],
    [
      NotificationTypeEnum.error,
      <div
        className="flex h-7 w-7 items-center justify-center rounded-[50%] bg-red-50"
        key="system"
      >
        <CloseCircleOutlined className="text-red-500" />
      </div>,
    ],
    [
      NotificationTypeEnum.information,
      <div
        className="flex h-7 w-7 items-center justify-center rounded-[50%] bg-slate-50"
        key="system"
      >
        <ExclamationCircleOutlined className="text-slate-500" />
      </div>,
    ],
    [
      NotificationTypeEnum.warn,
      <div
        className="flex h-7 w-7 items-center justify-center rounded-[50%] bg-yellow-50"
        key="system"
      >
        <WarningOutlined className="text-yellow-500" />
      </div>,
    ],
  ]);

  return typeCorrespondIconMap.get(type);
};
