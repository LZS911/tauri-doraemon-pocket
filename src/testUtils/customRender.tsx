/* eslint-disable @typescript-eslint/no-explicit-any */
import { render, renderHook } from '@testing-library/react';
import { Provider } from 'react-redux';
import {
  BrowserRouter,
  MemoryRouter,
  MemoryRouterProps,
  Router,
  RouterProps,
} from 'react-router-dom';
import { Dictionary } from '../typing/common.type';
import { storeFactory } from './mockRedux';
type RenderParams = Parameters<typeof render>;

export const renderWithRedux: (arg: any) => ReturnType<typeof render> = (
  ...[ui, option, initStore]: [...RenderParams, Dictionary?]
) => {
  return render(
    <Provider store={storeFactory(initStore)}>{ui}</Provider>,
    option
  );
};
export const renderWithRouter: any = (...[ui, option]: [...RenderParams]) => {
  return render(<BrowserRouter>{ui}</BrowserRouter>, option);
};

export const renderHooksWithRedux = <TProps, TResult>(
  hooks: (props: TProps) => TResult,
  storeState: Dictionary = {}
) => {
  return renderHook(hooks, {
    wrapper: Provider,
    initialProps: {
      store: storeFactory(storeState),
    },
  } as any);
};

export const renderWithServerRouter: any = (
  ...[ui, option, props]: [...RenderParams, RouterProps]
) => {
  return render(<Router {...props}>{ui}</Router>, option);
};

export const renderWithServerRouterAndRedux: any = (
  ...[ui, option, props]: [...RenderParams, RouterProps]
) => {
  return render(
    <Router {...props}>
      {' '}
      <Provider store={storeFactory()}>{ui}</Provider>,
    </Router>,
    option
  );
};

export const renderWithMemoryRouter: any = (
  ...[ui, option, props]: [...RenderParams, MemoryRouterProps?]
) => {
  return render(<MemoryRouter {...props}>{ui}</MemoryRouter>, option);
};
