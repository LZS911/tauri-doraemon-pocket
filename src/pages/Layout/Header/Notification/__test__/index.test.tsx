import { fireEvent, screen, waitFor } from '@testing-library/react';
import { renderWithRedux } from '../../../../../testUtils/customRender';
import Notification from '../index';

describe('test Notification', () => {
  beforeEach(() => {
    jest.useFakeTimers();
  });
  afterEach(() => {
    jest.useRealTimers();
  });

  test('should match snapshot', async () => {
    const { container } = renderWithRedux(<Notification />);

    await waitFor(() => {
      jest.advanceTimersByTime(0);
    });

    fireEvent.click(screen.getByTestId('popover-switch-icon'));

    expect(container).toMatchSnapshot();
  });

  test('should mark as all read when clicking icon', async () => {
    const { container } = renderWithRedux(<Notification />);
    await waitFor(() => {
      jest.advanceTimersByTime(0);
    });

    fireEvent.click(screen.getByTestId('popover-switch-icon'));

    // expect(container.querySelector('.ant-badge-count')).toHaveAttribute(
    //   'title',
    //   '2'
    // );

    fireEvent.click(screen.getByTestId('mark-all-read'));

    // expect(container.querySelector('.ant-badge-count')).not.toBeInTheDocument();
  });
});
