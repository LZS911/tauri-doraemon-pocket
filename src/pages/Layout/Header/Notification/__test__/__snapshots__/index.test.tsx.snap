// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test Notification should match snapshot 1`] = `
<div>
  <div
    class="hover:bg-slate-200 dark:hover:bg-slate-400 after:water-wave-hide active:after:water-wave-show mr-2 bg-slate-100 dark:bg-black ant-popover-open transparent relative flex cursor-pointer items-center justify-center rounded-md bg-transparent p-2 transition-colors dark:text-white"
    data-testid="popover-switch-icon"
  >
    <span
      class="ant-badge ant-badge-status leading-[0.875rem]"
    >
      <span
        class="dark:text-white"
      >
        <span
          aria-label="bell"
          class="anticon anticon-bell text-[1rem]"
          role="img"
        >
          <svg
            aria-hidden="true"
            data-icon="bell"
            fill="currentColor"
            focusable="false"
            height="1em"
            viewBox="64 64 896 896"
            width="1em"
          >
            <path
              d="M816 768h-24V428c0-141.1-104.3-257.7-240-277.1V112c0-22.1-17.9-40-40-40s-40 17.9-40 40v38.9c-135.7 19.4-240 136-240 277.1v340h-24c-17.7 0-32 14.3-32 32v32c0 4.4 3.6 8 8 8h216c0 61.8 50.2 112 112 112s112-50.2 112-112h216c4.4 0 8-3.6 8-8v-32c0-17.7-14.3-32-32-32zM512 888c-26.5 0-48-21.5-48-48h96c0 26.5-21.5 48-48 48zM304 768V428c0-55.6 21.6-107.8 60.9-147.1S456.4 220 512 220c55.6 0 107.8 21.6 147.1 60.9S720 372.4 720 428v340H304z"
            />
          </svg>
        </span>
      </span>
      <sup
        class="ant-scroll-number ant-badge-count ant-badge-count-sm"
        data-show="true"
        style="margin-top: 2px; right: 0px; background: rgb(24, 144, 255);"
        title="2"
      >
        <span
          class="ant-scroll-number-only"
          style="transition: none;"
        >
          <span
            class="ant-scroll-number-only-unit current"
          >
            2
          </span>
        </span>
      </sup>
    </span>
  </div>
</div>
`;
